observing/obsstate.py

```diff
@@ -191,7 +216,29 @@ def iterate_max_session_id():
         except TypeError:
             max_session_id = 0
 
-    return str(max_session_id+1)
+    return max_session_id+1
+
+
+def check_and_create_pi(pi_name):
+    """
+    Check if a PI exists in the database, and if not, create one.
+    """
+
+    with connection_factory() as conn:
+        c = conn.cursor()
+        c.execute("SELECT pi_id FROM pis WHERE pi_name = ?", (pi_name,))
+        row = c.fetchone()
+        if row:
+            return row[0]
+        else:
+            c.execute("SELECT MAX(pi_id) FROM pis")
+            try:
+                max_pi_id = int(c.fetchone()[0])
+            except TypeError:
+                max_pi_id = 0
+            new_pi_id = max_pi_id + 1
+            c.execute("INSERT INTO pis VALUES (?, ?)", (new_pi_id, pi_name))
+            return str(new_pi_id)
```

Comment:
this is me being paranoid. But consider moving the return statement outside the `with` block. The context manager may try to do extra work (and I think it does) and that work may throw exceptions. In that case one should throw before returning instead of the other way around.

Also I thought pi_id is an int?